### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by leoenge**

### CODE REVIEW

**Review by ajlace — COMMENTED** [review]
Looks amazing! Super great that you included the Google Collab link.

As Lab 2 is included in this PR, it will not get merged until Lab 2 is graded.

**Comment by ajlace** [comment]
Do not close the Pull Reqeust as we might miss it when grading. Closing it implies that it is not ready for review.

**Comment by johanhoffman** [comment]
Ok!

File: Lab-2/leoenge_lab2.ipynb [hunk 1]
```diff
@@ -0,0 +1,916 @@
...
+        "    A = np.array([[1,0,1],[0,1,1],[1,0,1]])\n",
+        "    with self.assertRaises(Exception):\n",
+        "      qr_factorization(A)\n",
+        "  \n",
+        "  def test_first_residual(self):\n",
+        "    count = 0\n",
+        "    while count < 100:\n",
+        "      dim = np.random.randint(1,10)\n",
+        "      A = np.random.rand(dim, dim)\n",
+        "      if np.linalg.det(M) == 0:\n",
```
Comment (ajlace) [inline Lab-2/leoenge_lab2.ipynb]: I am assuming that this is supposed to be A instead?

File: Lab-2/leoenge_lab2.ipynb [hunk 2]
```diff
@@ -0,0 +1,916 @@
...
+        "      #Testing for which decimal value it fails to equal. Do not however care about more than 7 decimals.\n",
+        "      for i in range(3,7):\n",
+        "        np.testing.assert_almost_equal(A.dot(x), b, decimal=i)\n",
+        "  \n",
+        "  def test_reversed_residual(self):\n",
+        "    count = 0\n",
+        "    while count < 100:\n",
+        "      dim = np.random.randint(1,10)\n",
+        "      A = np.random.rand(dim, dim)\n",
+        "      if np.linalg.det(M) == 0:\n",
```
Comment (ajlace) [inline Lab-2/leoenge_lab2.ipynb]: This should also be A.

File: Lab-2/leoenge_lab2.ipynb [hunk 3]
```diff
@@ -0,0 +1,916 @@
...
+        "colab_type": "code",
+        "outputId": "a351a589-f45e-4149-83eb-06799f304f8d",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 238
+        }
+      },
+      "source": [
+        "class Test(unittest.TestCase):\n",
+        "  def test_singular(self):\n",
```
Comment (ajlace) [inline Lab-2/leoenge_lab2.ipynb:497]: This test seems to fail no matter what changes I make. Look through it a second time to confirm it behaves the way you want it to and make updates accordingly.

**Review by ajlace — DISMISSED** [review]
Do correct the comments I have given, however other than that the lab looks good.

**Pull request merged**